model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java

```diff
@@ -163,24 +220,8 @@ private static void updateAuthSession(RootAuthenticationSessionEntity rootAuthSe
                 }
             }
         }
-    }
-
-    private void lazyInit(KeycloakSession session) {
-        if (authSessionsCache == null) {
-            synchronized (this) {
-                if (authSessionsCache == null) {
-                    InfinispanConnectionProvider connections = session.getProvider(InfinispanConnectionProvider.class);
-                    authSessionsCache = connections.getCache(InfinispanConnectionProvider.AUTHENTICATION_SESSIONS_CACHE_NAME);
-
-                    keyGenerator = new InfinispanKeyGenerator();
 
-                    ClusterProvider cluster = session.getProvider(ClusterProvider.class);
-                    cluster.registerListener(AUTHENTICATION_SESSION_EVENTS, this::updateAuthNotes);
-
-                    log.debugf("[%s] Registered cluster listeners", authSessionsCache.getCacheManager().getAddress());
-                }
-            }
-        }
+        this.authSessionsCache.replace(rootAuthSession.getId(), new SessionEntityWrapper<>(rootAuthSessionWrapper.getLocalMetadata(), rootAuthSession));
```

Comment:
@pruivo Can this work?

Reply:
Yes and no. I don't fully understand this method but it is invoked from an event notification. All KC nodes in the cluster will invoke replace at the same time... it works but it is not good for performance reasons :)